doc/designs/quic-design/server/quic-polling.md

```diff
@@ -0,0 +1,1637 @@
+specifying a SSL object pointer:
+
+```c
+#define BIO_POLL_DESCRIPTOR_SSL 2   /* (SSL *) */
+
+typedef struct bio_poll_descriptor_st {
+    int type;
+    union {
+        ...
+        SSL     *ssl;
+    } value;
+} BIO_POLL_DESCRIPTOR;
+```
+
+Event Types and Representation
+------------------------------
+
+Regardless of the API design chosen, event types can first be defined:
+
+```c
+#define OSSL_POLL_EVENT_NONE       0
+
+/* stream/default stream readable or reset */
+#define OSSL_POLL_EVENT_R          (1U << 0)
+
+/* stream/default stream writable or stopped */
+#define OSSL_POLL_EVENT_W          (1U << 1)
+
+/* error (i.e. connection terminating) */
+#define OSSL_POLL_EVENT_E          (1U << 2)
+
+/* incoming bidi stream */
+#define OSSL_POLL_EVENT_ISB        (1U << 3)
+
+/* incoming uni stream */
+#define OSSL_POLL_EVENT_ISU        (1U << 4)
+
+/* incoming connection */
+#define OSSL_POLL_EVENT_IC         (1U << 5)
+
+/* can create new outgoing bidi stream */
+#define OSSL_POLL_EVENT_OSB        (1U << 6)
+
+/* can create new outgoing uni stream */
+#define OSSL_POLL_EVENT_OSU        (1U << 7)
+
+#define OSSL_POLL_EVENT_RW         (OSSL_POLL_EVENT_R | OSSL_POLL_EVENT_W)
+#define OSSL_POLL_EVENT_RWE        (OSSL_POLL_EVENT_RW | OSSL_POLL_EVENT_E)
+#define OSSL_POLL_EVENT_IS         (OSSL_POLL_EVENT_ISB | OSSL_POLL_EVENT_ISU)
+#define OSSL_POLL_EVENT_I          (OSSL_POLL_EVENT_IS | OSSL_POLL_EVENT_IC)
+#define OSSL_POLL_EVENT_OS         (OSSL_POLL_EVENT_OSB | OSSL_POLL_EVENT_OSU)
+```
+
+Designs
+-------
+
+Two designs are considered here:
+
+- Sketch A: An “immediate-mode” poller interface similar to poll(2).
+
+- Sketch B: A “registered” poller interface similar to BSD's kqueue(2) (or Linux's
+  epoll(2)).
+
+Sketch A is simpler but is likely to be less performant. Sketch B is a bit more
+elaborate but can offer more performance. It is possible to offer both APIs if
+desired.
+
+### Sketch A: One-Shot/Immediate Mode API
+
+We define a common structure for representing polled events:
+
+```c
+typedef struct ossl_poll_item_st {
+    BIO_POLL_DESCRIPTOR desc;
+    uint64_t            events, revents;
+} OSSL_POLL_ITEM;
+```
+
+This structure works similarly to the `struct pollfd` structure used by poll(2).
+`desc` describes the object to be polled, `events` is a bitmask of
+`OSSL_POLL_EVENT` values describing what events to listen for, and `revents` is
+a bitmask of zero or more events which are actually raised.
+
+Polling implementations are only permitted to modify the `revents` field in a
+`OSSL_POLL_ITEM` structure passed by the caller.
+
+```c
+/*
+ * SSL_poll
+ * --------
+ *
+ * SSL_poll evaluates each of the items in the given array of OSSL_POLL_ITEMs
+ * and determines which poll items have relevant readiness events raised. It is
+ * similar to POSIX poll(2).
+ *
+ * The events field of each item specifies the events the caller is interested
+ * in and is the sum of zero or more OSSL_POLL_EVENT_* values. When using
```

Comment:
Not necessarily. You can register a descriptor and not enable any events on it, because enabling an event on an already registered descriptor is likely to be more efficient than registering a new descriptor.

This is actually likely to be a common case, namely due to `OSSL_POLL_EVENT_FLAG_DISPATCH`. In this feature an event is disabled after it is dispatched, which is necessary to avoid the thundering herd problem.

So think of the following happening repeatedly at a very high frequency:

- Application polls
- A registered descriptor with only the 'read' event enabled is ready, dispatch mode is enabled, so this is disabled (leaving no events enabled) and the application is informed
- Application polls again and asks for the 'read' event to be reenabled in a single call

So it is actually quite likely to have a registered descriptor that transitions between having no events enabled, and some events enabled, at high frequency. This makes optimising this case quite desirable.

In any case it is just an emergent consequence of the design and doesn't cost us anything to support.